pkg/util/net.go

```diff
@@ -0,0 +1,134 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"bufio"
+	"bytes"
+	"net"
+	"os/exec"
+	"strings"
+)
+
+type NetController interface {
+	GetInterfaceAddresses(ifName string) [](net.IPNet)
+	GetNMDeviceState() (*bytes.Buffer, error)
+}
+
+func NewNetController() NetController {
+	return &unixNet{}
+}
+
+var _ NetController = (*unixNet)(nil)
+
+type unixNet struct{}
+
+func (un *unixNet) GetInterfaceAddresses(ifName string) (res []net.IPNet) {
+	var (
+		iface *net.Interface
+		addrs []net.Addr
+		err   error
+	)
+	res = []net.IPNet{}
+
+	if iface, err = net.InterfaceByName(ifName); err != nil {
+		return
+	}
+	if addrs, err = iface.Addrs(); err != nil {
+		return
+	}
+	for _, addr := range addrs {
+		if ipNet, ok := addr.(*net.IPNet); ok {
+			res = append(res, *ipNet)
+		}
+	}
+	return
+}
+
+func (un *unixNet) GetNMDeviceState() (buf *bytes.Buffer, err error) {
+	buf = &bytes.Buffer{}
+	cmd := exec.Command("nmcli", "-g", "DEVICE,STATE", "device")
+	cmd.Stdout = buf
+
+	err = cmd.Run()
+
+	return
+}
+
+// GetIPByIfName returns a string containing the IP address of the network
+// interface name (ifName) passed as argument.
+// The returned IP address is the IPv4 one but the IPv6 one is returned if
+// the IPv4 address is not available.
+// This function cannot fail: if no address can be retrieved an empty string
+// is returned.
+func GetIPByIfName(ifName string) string {
+	return getIPByIfName(ifName, NewNetController())
+}
+
+func getIPByIfName(ifName string, nc NetController) string {
+	var (
+		ips  []net.IPNet
+		ipv6 net.IP
+	)
+	if ips = nc.GetInterfaceAddresses(ifName); len(ips) == 0 {
+		return ""
+	}
+	for _, ip := range ips {
+		if ipv4 := ip.IP.To4(); ipv4 != nil {
+			return ipv4.String()
+		}
+
+		// Let's keep an IPv6 address as fallback if the interface has no IPv4 addresses
+		if ipv6 == nil {
+			ipv6 = ip.IP.To16()
+		}
+	}
+	if ipv6 != nil {
+		return ipv6.String()
+	}
+	return ""
+}
+
+// GetNMActivatedDevices() returns a slice with the names of the network interfaces
+// that have an active NetworkManager connection profile managed by NetworkManager.
+// Assumed connection profiles are not counted in.
+// This function cannot fail: if NM data cannot be inspected an empty slice is returned.
+func GetNMActivatedDevices() (nmDevs []string) {
+	return getNMActivatedDevices(NewNetController())
+}
+
+func getNMActivatedDevices(nc NetController) (nmDevs []string) {
+	var buf *bytes.Buffer
+	var err error
+	nmDevs = []string{}
+
+	if buf, err = nc.GetNMDeviceState(); err != nil {
+		return
```

Comment:
Would be nice to print the error so that it does not fail silently, for example if `nmcli` is missing/broken.